third_party/txt/src/txt/fuchsia_font_manager.cc

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "txt/fuchsia_font_manager.h"
+
+#include <zx/vmar.h>
+
+#include "lib/fsl/vmo/sized_vmo.h"
+#include "lib/fxl/logging.h"
+#include "txt/asset_font_style_set.h"
+
+namespace txt {
+
+namespace {
+
+fonts::FontProviderPtr* g_font_provider = nullptr;
+
+fonts::FontProviderPtr& GetFontProvider() {
+  FXL_CHECK(g_font_provider);
+  return *g_font_provider;
+}
+
+void UnmapMemory(const void* buffer, void* context) {
+  static_assert(sizeof(void*) == sizeof(uint64_t), "pointers aren't 64-bit");
+  const uint64_t size = reinterpret_cast<uint64_t>(context);
+  zx::vmar::root_self().unmap(reinterpret_cast<uintptr_t>(buffer), size);
+}
+
+sk_sp<SkData> MakeSkDataFromVMO(const fsl::SizedVmoTransportPtr& vmo) {
+  if (!fsl::SizedVmo::IsSizeValid(vmo->vmo, vmo->size) ||
+      vmo->size > std::numeric_limits<size_t>::max()) {
+    return nullptr;
+  }
+  uint64_t size = vmo->size;
+  uintptr_t buffer = 0;
+  zx_status_t status = zx::vmar::root_self().map(0, vmo->vmo, 0, size,
+                                                 ZX_VM_FLAG_PERM_READ, &buffer);
+  if (status != ZX_OK)
+    return nullptr;
+  return SkData::MakeWithProc(reinterpret_cast<void*>(buffer), size,
+                              UnmapMemory, reinterpret_cast<void*>(size));
+}
+
+fonts::FontSlant ToFontSlant(SkFontStyle::Slant slant) {
+  return (slant == SkFontStyle::kItalic_Slant) ? fonts::FontSlant::ITALIC
+      : fonts::FontSlant::UPRIGHT;
+}
+
+}  // anonymous namespace
+
+FuchsiaFontManager::FuchsiaFontManager() = default;
+
+FuchsiaFontManager::~FuchsiaFontManager() = default;
+
+void FuchsiaFontManager::SetFontProvider(fonts::FontProviderPtr provider) {
```

Comment:
If the font manager is moved into the content handler, we wont need a singleton in this library either. The font manager instance can then be managed by the embedder.